### PR TITLE
[TMP] TIP-452: Remove FilteredProductReader to use ProductReader

### PR DESCRIPTION
**Pull request by pierallard**
# Do not review this PR. Please review https://github.com/akeneo/pim-community-dev/pull/4745

In order to simplify the import/export stack, we found that FilteredProductReader and ProductReader were very close. Indeed, the 2 classes read options from stepConfiguration, create a PQB, and read items one by one, and have exactly the same signature.

This PR replace the FilteredProductReader with ProductReader. There is no drastic changes, more cleaning than real changes.

This PR normalize the input filters for import/export and for product query commands to, using a new ProductQueryBuilderFactory. This new factory creates a PQB with an input filters list, and it's used on import/export/commandline (no remaining other filters analysis like this).

The "standard" format for filters is now:
`[field: xxx, operator: xxx, value: xxx, context: [locale:xxx, scope: xxx]]`

FAQ
- why did you remove the object detacher here ? no risk of memory leaks ?
Excellent question. The thing is that after the read process, every product is "manipulated" and need to be re-hydrated. If object is detached in the reading part, there is 2 hydrations of the same object for nothing. In the next steps (e.g. ProductProcessor or ProductWriter), there is an ObjectDetacher after the process. So, no changes of memory if we remove this.

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | y
| Added Behats                      | n
| Changelog updated                 | y
| Review and 2 GTM                  | TODO
| Migration script                  | n
| Tech Doc                          | n

### CODE REVIEW

File: src/Pim/Component/Connector/Reader/Database/ProductReader.php [hunk 1]
```diff
@@ -38,9 +40,6 @@ class ProductReader extends AbstractConfigurableStepElement implements ItemReade
     /** @var MetricConverter */
     protected $metricConverter;
 
-    /** @var ObjectDetacherInterface */
-    protected $objectDetacher;
```
Comment (solivier) [inline src/Pim/Component/Connector/Reader/Database/ProductReader.php]: why did you remove the object detacher here ? no risk of memory leaks ?

File: src/Pim/Component/Connector/Reader/Database/ProductReader.php [hunk 2]
```diff
@@ -108,16 +91,17 @@ public function read()
...
             $product = $this->products->current();
             $this->stepExecution->incrementSummaryInfo('read');
             $this->products->next();
         }
 
         if (null !== $product) {
-            $this->objectDetacher->detach($product);
             $channel = $this->getConfiguredChannel();
-            $this->metricConverter->convert($product, $channel);
+            if ($channel) {
```
Comment (solivier) [inline src/Pim/Component/Connector/Reader/Database/ProductReader.php]: channel is not a boolean use null !== $channel instead

File: src/Pim/Bundle/EnrichBundle/Connector/Reader/MassEdit/FilteredVariantGroupProductReader.php [hunk 3]
```diff
@@ -3,57 +3,74 @@
...
+    protected $cleanedFilters = null;
 
     /**
      * @param ProductQueryBuilderFactoryInterface $pqbFactory
      * @param VariantGroupCleaner                 $cleaner
      */
-    public function __construct(ProductQueryBuilderFactoryInterface $pqbFactory, VariantGroupCleaner $cleaner)
-    {
-        parent::__construct($pqbFactory);
+    public function __construct(
```
Comment (solivier) [inline src/Pim/Bundle/EnrichBundle/Connector/Reader/MassEdit/FilteredVariantGroupProductReader.php]: why not using composition here ?

File: src/Pim/Component/Connector/Reader/Database/ProductReader.php (hunk 1 above)
Comment (pierallard) [inline src/Pim/Component/Connector/Reader/Database/ProductReader.php]: Excellent question. The thing is that after the read process, every product is "manipulated" and need to be re-hydrated. If object is detached in the reading part, there is 2 hydrations of the same object for nothing. In the next steps (e.g. ProductProcessor), there is an ObjectDetacher after the process. So, no changes of memory if we remove this.

File: src/Pim/Bundle/EnrichBundle/Connector/Reader/MassEdit/FilteredVariantGroupProductReader.php [hunk 4]
```diff
@@ -3,57 +3,74 @@
...
  */
-class FilteredVariantGroupProductReader extends FilteredProductReader
+class FilteredVariantGroupProductReader extends ProductReader
 {
     /** @var VariantGroupCleaner */
     protected $cleaner;
 
     /** @var array */
-    protected $cleanedFilters;
+    protected $cleanedFilters = null;
```
Comment (jjanvier) [inline src/Pim/Bundle/EnrichBundle/Connector/Reader/MassEdit/FilteredVariantGroupProductReader.php]: useless, it's by default in PHP

File: src/Pim/Component/Connector/Reader/Database/ProductReader.php [hunk 5]
```diff
@@ -132,20 +116,90 @@ public function setStepExecution(StepExecution $stepExecution)
...
             throw new ObjectNotFoundException(sprintf('Channel with "%s" code does not exist', $channelCode));
         }
 
         return $channel;
     }
+
+    /**
+     * @return ProductQueryBuilderInterface
+     */
+    protected function getProductQueryBuilder()
```
Comment (jjanvier) [inline src/Pim/Component/Connector/Reader/Database/ProductReader.php]: createProductQueryBuilder

**Comment by solivier** [comment]
gtm
